notebooks/analyze_looping_across_lending_protocols.ipynb

```diff
@@ -713,44 +637,8 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "\n",
-    "# Helper function to parse token amounts\n",
-    "def parse_token_amounts(token_str):\n",
-    "    tokens = []\n",
-    "    if token_str:\n",
-    "        token_pairs = token_str.split(', ')\n",
-    "        for token_pair in token_pairs:\n",
-    "            parts = token_pair.split(': ')\n",
-    "            if len(parts) == 2:\n",
-    "                token_name, amount = parts\n",
-    "                tokens.append((token_name, float(amount)))\n",
-    "    return tokens\n",
-    "\n",
-    "# Function to extract user actions from the dataframes\n",
-    "def extract_user_actions(dataframes, tokens):\n",
-    "    user_actions = defaultdict(list)\n",
-    "    for protocol, df in dataframes.items():\n",
-    "        for _, row in df.iterrows():\n",
-    "            user = row['User']\n",
-    "            collateral_tokens = parse_token_amounts(row['Collateral'])\n",
-    "            debt_tokens = parse_token_amounts(row['Debt'])\n",
-    "\n",
-    "            # Store collateral actions\n",
-    "            for col_token, col_amount in collateral_tokens:\n",
-    "                if col_token in tokens:\n",
-    "                    user_actions[user].append(('collateral', col_token, col_amount, protocol))\n",
-    "\n",
-    "            # Store debt actions\n",
-    "            for debt_token, debt_amount in debt_tokens:\n",
-    "                if debt_token in tokens:\n",
-    "                    user_actions[user].append(('debt', debt_token, debt_amount, protocol))\n",
-    "    return user_actions\n",
-    "\n",
-    "# Extract user actions\n",
-    "user_actions = extract_user_actions(dataframes, tokens)\n",
-    "\n",
     "# Function to count token usage in collateral and debt\n",
-    "def count_token_usage(user_actions):\n",
+    "def count_token_usage(user_actions: DefaultDict[str, List[Tuple[str, str, float, str]]]) -> Tuple[DefaultDict[str, int], DefaultDict[str, int]]:\n",
```

Comment:
Pls run the code before you send the PR back. You would easily see that it produces a bunch of `NameErrors` 😉

Reply:
Done! ✅